### PR TITLE
Heap Implementation

### PR DESCRIPTION
**Pull request by ChinthanBD**

### CODE REVIEW

File: heap.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    def __init__(self):
+        self.array = DynamicArray(10)
+
+    def add(self, value):
+        self.array.append(value)
+        self.heapify()
+
+    def delete(self):
+        if self.array.is_empty():
+            return False
```
Comment (tejas-k3) [inline heap.py:14]: Interesting choice! More appropriate way to do would be to throw error as you're trying to delete from EMPTY. Which approach sounds more appropriate to you  @ChinthanBD ?

File: heap.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+    def heapify(self):
+        child_index = self.array.fetch_size() - 1
+
+        while child_index > 0:
+            parent_index = (child_index - 1) // 2
+            child_element = self.array.get(child_index)
+            parent_element = self.array.get(parent_index)
+
+            if child_element is not None and parent_element is not None and child_element < parent_element:
```
Comment (tejas-k3) [inline heap.py:32]: This can be simplified as `if x is not None` is same as `if x`. Pls simplify

File: heap.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    def add(self, value):
+        self.array.append(value)
+        self.heapify()
+
+    def delete(self):
+        if self.array.is_empty():
+            return False
+        self.swap_elements(0, self.array.fetch_size()-1)
+        self.array.current_index -= 1
+        self.array.size -= 1
```
Comment (tejas-k3) [inline heap.py]: Why directly manipulating attributes of internal data structure ? You need to make use of APIs (Methods declared within that data structure) OR Create attribute for your current class

File: heap.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        self.array = DynamicArray(10)
+
+    def add(self, value):
+        self.array.append(value)
+        self.heapify()
+
+    def delete(self):
+        if self.array.is_empty():
+            return False
+        self.swap_elements(0, self.array.fetch_size()-1)
```
Comment (tejas-k3) [inline heap.py]: After swap_elements, there's no delete operation happening, we should add the below line :
```
self.array.delete(self.array.fetch_size()-1)
```

**Review by tejas-k3 — CHANGES_REQUESTED** [review]

**Comment by ChinthanBD** [comment]
Code output

```
C:/Users/cdevaraju/Desktop/123/code/CustomDS/heap.py
Min Heap after adding elements: [1, 3, 8, 5, 7, None, None, None, None, None]
Element deleted successfully: True
Min Heap after deleting element: [7, 3, 8, 5, None, None, None, None, None, None]
Element deleted successfully: True
Min Heap after attempting to delete from an empty heap: [5, 3, 8, None, None, None, None, None, None, None]

Process finished with exit code 0
```

File: heap.py (hunk 4 above)
Comment (ChinthanBD) [reply heap.py]: changed delete method

File: heap.py (hunk 3 above)
Comment (ChinthanBD) [reply heap.py]: Done

File: heap.py (hunk 2 above)
Comment (ChinthanBD) [reply heap.py:32]: Done

File: heap.py (hunk 1 above)
Comment (ChinthanBD) [reply heap.py:14]: Don't want to throw error. Exiting gracefully returning false seems good

**Comment by tejas-k3** [comment]
Check your output after 2nd delete.. min heap aint min heapping

**Comment by ChinthanBD** [comment]
Code output:

```
C:\Users\cdevaraju\AppData\Local\Programs\Python\Python39\python.exe C:/Users/cdevaraju/Desktop/123/code/CustomDS/heap.py
Min Heap after adding elements: [1, 3, 8, 5, 7]
Element deleted successfully: True
Min Heap after deleting element: [7, 3, 8, 5]
Element deleted successfully: True
Min Heap after attempting to delete from an empty heap: [5, 3, 8]

Process finished with exit code 0
```

**Comment by tejas-k3** [comment]
Good fix for `None` value coming while printing the MinHeap

**Comment by ChinthanBD** [comment]
Code output:

```
Min Heap after adding elements: [1, 3, 8, 5, 7]
Element deleted successfully: True
Min Heap after deleting element: [3, 5, 8, 7]
Element deleted successfully: True
Min Heap after attempting to delete from an empty heap: [5, 7, 8]

Process finished with exit code 0
```

**Comment by tejas-k3** [comment]
Pls raise a new MR containing ticket raised for issue.

**Pull request merged**